core/src/main/java/org/apache/druid/java/util/common/Cleaners.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.java.util.common;
+
+import org.apache.druid.utils.JvmUtils;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+
+public class Cleaners
+{
+  public interface Cleanable
+  {
+    void clean();
+  }
+
+  public interface Cleaner
+  {
+    Cleanable register(Object object, Runnable runnable);
+  }
+
+  private static final Cleaner CLEANER;
+  private static final RuntimeException CLEANER_NOT_SUPPORTED_EXCEPTION;
+
+  static {
+    Cleaners.Cleaner cleaner = null;
+    RuntimeException exception = null;
+    try {
+      cleaner = takeMeToTheCleaners();
+    }
+    catch (RuntimeException e) {
```

Comment:
Is there a chance that an unchecked exception here might be an `Error` rather than a `RuntimeException`? If so would be better to use `Throwable`.

Reply:
The main class of error I'd be worried about would be linkage errors, and from what I can tell MethodHandles take care of catching those and wrapping them when resolving methods. Other than that invoke() and invokeExact could also throw linkage errors, but we already have to catch any Throwables resulting from those.